### PR TITLE
cross platform support

### PR DESCRIPTION
**Pull request by xujihui1985**
## Reason for This PR

cross platform support

## Description of Changes

use mio instead of epoll to support cross platform

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: src/connection.rs [hunk 1]
```diff
@@ -4,16 +4,16 @@
 use std::collections::VecDeque;
 use std::fs::File;
 use std::io::{Read, Write};
-use std::os::unix::io::FromRawFd;
+use std::os::unix::prelude::{AsRawFd, FromRawFd, RawFd};
```
Comment (jiangliu) [inline src/connection.rs]: Should this be `std::os::unix::{AsRawFd, FromRawFd, RawFd};`?

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+on: [push]
+
+name: CI
+
```
Comment (jiangliu) [inline .github/workflows/rust.yml]: Could this file be merged into the existing CI workflow?

File: src/connection.rs [hunk 3]
```diff
@@ -61,6 +61,16 @@ pub struct HttpConnection<T> {
...
+impl<T> AsRawFd for HttpConnection<T>
+where
+    T: AsRawFd,
+{
+    fn as_raw_fd(&self) -> RawFd {
+        self.stream.as_raw_fd()
+    }
+}
+
+// impl<T: Read + Write + ScmSocket> HttpConnection<T> {
```
Comment (jiangliu) [inline src/connection.rs:73]: This line should be removed.

File: src/connection.rs [hunk 4]
```diff
@@ -172,17 +182,23 @@ impl<T: Read + Write + ScmSocket> HttpConnection<T> {
...
         }];
 
-        // Safe because we have mutably borrowed buf and it's safe to write
-        // arbitrary data to a slice.
         let (read_count, fd_count) = unsafe {
             self.stream
                 .recv_with_fds(&mut iovecs, &mut fds)
                 .map_err(ConnectionError::StreamReadError)?
         };
+        // let mut iovecs = [IoSliceMut::new(buf)];
```
Comment (jiangliu) [inline src/connection.rs]: Should these be removed?

File: src/server.rs [hunk 5]
```diff
@@ -600,39 +635,53 @@ impl HttpServer {
...
-            .ctl(
-                epoll::ControlOperation::Delete,
-                stream_fd,
-                epoll::EpollEvent::new(epoll::EventSet::IN, stream_fd as u64),
-            )
+    fn epoll_del(poll: &Poll, stream_fd: RawFd) -> Result<()> {
+        poll.registry()
+            .deregister(&mut SourceFd(&stream_fd))
             .map_err(ServerError::IOError)
+        // epoll
```
Comment (jiangliu) [inline src/server.rs]: Should these be removed?

File: src/connection.rs (hunk 3 above)
Comment (xujihui1985) [reply src/connection.rs:73]: yes, all the commented code should be cleaned up

File: src/response.rs [hunk 6]
```diff
@@ -151,6 +151,7 @@ impl ResponseHeaders {
         self.write_allow_header(buf)?;
         self.write_deprecation_header(buf)?;
 
+
```
Comment (jiangliu) [inline src/response.rs]: Unnecessary change?

Comment (xujihui1985) [reply src/response.rs]: ya, I write the content_length to response header no matter content_length is 0, but after I check the http spec carefully, I found if the status code is 1xx or 204, we must not set the content length header, so I revert the change. However, we may need to consider the case when http status code is 200 and content_length is zero, in this case content length should be set to the response header

File: .github/workflows/rust.yml (hunk 2 above)
Comment (xujihui1985) [reply .github/workflows/rust.yml]: this can be removed as there is a better CI workflow in upstream

**Review by jiangliu — APPROVED** [review]
